lib/SIL/OwnershipUtils.cpp

```diff
@@ -438,3 +387,128 @@ bool BorrowScopeIntroducingValue::visitLocalScopeTransitiveEndingUses(
 
   return foundError;
 }
+
+//===----------------------------------------------------------------------===//
+// Searching for BorrowScopeIntroducers in Def-Use Graph
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+/// A generator of borrow scope introducing values.
+struct BorrowScopeIntroducingValueGenerator {
+  SmallVector<SILValue, 32> worklist;
+
+  BorrowScopeIntroducingValueGenerator(SILValue value) {
+    assert(value.getOwnershipKind() == ValueOwnershipKind::Guaranteed);
+    worklist.push_back(value);
+  }
+
+  class Result {
+    Optional<BorrowScopeIntroducingValue> value;
+    bool hadError;
+
+  public:
+    Result(BorrowScopeIntroducingValue value) : value(value), hadError(false) {}
+    Result(llvm::NoneType) : value(), hadError(true) {}
+    Result() : value(), hadError(false) {}
+
+    bool finishedDueToError() const { return !value.hasValue() && hadError; }
+    bool finishedDueToNormalCompletion() const {
+      return !value.hasValue() && !hadError;
+    }
+
+    bool hasValue() const { return value.hasValue(); }
+    BorrowScopeIntroducingValue getValue() const { return value.getValue(); }
+  };
+
+  Result next();
+};
+
+} // end anonymous namespace
+
+BorrowScopeIntroducingValueGenerator::Result
+BorrowScopeIntroducingValueGenerator::next() {
+  while (!worklist.empty()) {
```

Comment:
I don't like the idea of a worklist here because it contradicts the basic design of a single borrow introducer for each scope.

Reply:
What if I have a struct literal that has multiple guaranteed parameters?

Reply:
In such a case, I think we would need to return both borrow introducers.

Reply:
Ok, if we allow a borrowed tuple or struct to be composed from borrowed elements, then this utility makes sense. The only thing I don't like is the general code that places all operands on the worklist. I feel like forwarding instructions (casts) should have an API that returns the single forwarded operand. And tuple/struct should be handled explicitly as the only thing that can "merge" borrow scopes.

There should be no reason to allow cond_br to handle borrowed values. No passes require critical edges in the mandatory pipeline. If it's done temporarily until we get around to enforcing that, then it should be a FIXME.